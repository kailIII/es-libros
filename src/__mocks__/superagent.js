let retainedCallback = undefined
const superagent = jest.genMockFromModule('superagent')


function get(/*url*/) {
  return {
    set: () => { return get() },
    send: () => { return get() },
    then: (successCallback, failiureCallback) => {
      retainedCallback = Object.create(null)
      retainedCallback.nextSuccess = successCallback
      retainedCallback.nextFailiure = failiureCallback
    }
  }
}

function __execSuccess(resp) {
  const nextSuccess = retainedCallback.nextSuccess
  if (nextSuccess)
    nextSuccess(resp)
}

function __execFailiure(error) {
  const nextFailiure = retainedCallback.nextFailiure
  if (nextFailiure)
    nextFailiure(resp)
}

superagent.get = get
superagent.__execSuccess = __execSuccess
superagent.__execFailiure = __execFailiure
module.exports = superagent
