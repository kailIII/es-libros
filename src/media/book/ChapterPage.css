div.container-markdown {
  padding: 14px;
  max-width: 700px;
  margin-left: auto;
  margin-right: auto;
}
h2.book-markdown {
  margin-top: 40px;
  text-align: center;
}

h3.book-markdown {
  font-family: 'ITC Charter';
  font-size: 22px;
  margin-top: 20px;
  margin-bottom: 20px;
  text-rendering: optimizeLegibility;
  border-bottom:1px solid #CCC;
  padding:0 6px 3px 6px;
}

h4.book-markdown {
  font-family: 'ITC Charter';
  font-size: 21px;
  margin-top: 20px;
  margin-bottom: 20px;
  text-rendering: optimizeLegibility;
  padding-bottom:3px;
}

ul.book-markdown {
  font-family: 'ITC Charter';
  font-size: 21px;
  line-height: 33.1833px;
  text-rendering: optimizeLegibility;
  list-style-position: inside;
  list-style-type: square;
}

p.book-markdown {
  font-family: 'ITC Charter';
  font-size: 21px;
  font-weight: 400;
  letter-spacing: -0.0666667px;
  line-height: 33.1833px;
  text-rendering: optimizeLegibility;
  text-align: left;
  text-indent: 20px;
}

code.book-markdown {
  font-family: 'ITC Charter';
  overflow-y: auto;
  display: block;
}

pre.book-markdown {
  white-space: pre-wrap;       /* css-3 */
  white-space: -moz-pre-wrap;  /* Mozilla, since 1999 */
  white-space: -pre-wrap;      /* Opera 4-6 */
  white-space: -o-pre-wrap;    /* Opera 7 */
  word-wrap: break-word;       /* Internet Explorer 5.5+ */
  font-family: 'ITC Charter';
  font-style: italic;

  display: block;
  font-size: 16px;
  margin: auto;
  text-align: left;
}

.next-chapter {
    background-color: #4CAF50; /* Green */
    border: none;
    color: white;
    padding: 15px 0px;
    width: 100%;
    display: block;
    text-align: center;
    text-decoration: none;
    font-size: 16px;
    margin: auto;
    cursor: pointer;
    overflow: hidden;
}
