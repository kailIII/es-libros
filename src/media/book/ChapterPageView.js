import React, { Component } from 'react';
import { Link } from 'react-router-dom'
import PropTypes from 'prop-types'

import markdownToComponentArray from '../../markdown'
import ProgressWheel from '../../comp/ProgressWheel'

const NextChapterButton = (props) => {
  const nextChapterUrl  = props.nextChapterUrl
  if (nextChapterUrl)
    return (
      <Link className="next-chapter" to={nextChapterUrl}>
        Siguiente Capitulo
      </Link>
    )
  return null
}

const ChapterPageContent = (props) => {
  const {
      text,
      title,
      nextChapterUrl,
    } = props

    return (
      <div>
        <div className="container-markdown">
          <h2 className="book-markdown">{title}</h2>
          { markdownToComponentArray(text, 'book-markdown') }
        </div>
        <NextChapterButton nextChapterUrl={nextChapterUrl} />
      </div>
    )
}

export default class ChapterPageView extends Component {

  render() {
    const {
      pageData
    } = this.props

    if (pageData)
      return <ChapterPageContent {...pageData} />
    else
      return <ProgressWheel />
  }
}

ChapterPageView.propTypes = {
  pageData: PropTypes.shape({
    nextChapterUrl: PropTypes.string,
    text: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
  })
}
