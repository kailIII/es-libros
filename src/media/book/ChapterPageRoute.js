import React from 'react';
import { Route } from 'react-router-dom'
import ChapterPage from './ChapterPage.js'

import { getChapterByIndex } from '../../server_data/PreloadedStateQueries.js'

const ChapterPageWrapper = (props) =>
  <ChapterPage getChapterByIndex={getChapterByIndex} {...props} />

const ChapterPageRoute = () =>
  <Route path="/book/:bookId/:chapterIndex" render={ChapterPageWrapper} />

export default ChapterPageRoute
