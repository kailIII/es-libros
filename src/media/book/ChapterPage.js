import React, { Component } from 'react';
import PropTypes from 'prop-types'

import { readChapter, unauthorizedResponseHandler } from '../../api'
import { findBookmark } from '../../storage/storage.js'
import { scrollByFraction } from '../../dom/Scroll'
import ChapterPageView from './ChapterPageView.js'

import './ChapterPage.css'

class ChapterPage extends Component {

  constructor(props) {
    super(props)
    this.state = {
      chapterText: '',
      bookmark: undefined,
    }
  }

  getNextChapterUrl = (bookId, i) => {
    const getChapterByIndex = this.props.getChapterByIndex
    const newIndex = parseInt(i, 10) + 1
    if (getChapterByIndex(bookId, newIndex))
      return `/book/${bookId}/${newIndex}`
  }

  pageWillChange = (nextProps) => {
    const {
      bookId,
      chapterIndex,
    } = this.props.match.params
    const {
      nextBookId,
      nextChapterIndex,
    } = nextProps.match.params

    return nextBookId !== bookId || nextChapterIndex !== chapterIndex
  }

  handleFetchChapterResponse = (props) => {
    return (resp) => {
      const bookId = props.match.params.bookId
      const search = props.location.search
      const chapterText = resp.text
      const nextState = { chapterText, bookmark: undefined }
      if (search && search.includes('bookmark=1')) {
        nextState.bookmark = findBookmark(bookId)
      }
      this.setState(nextState)
    }
  }

  fetchChapter = (props) => {
    const {
      bookId,
      chapterIndex,
    } = props.match.params

    readChapter(bookId, chapterIndex)
      .send()
      .then(this.handleFetchChapterResponse(props), unauthorizedResponseHandler())
  }

  mapStateToViewProps = () => {
    const {
      getChapterByIndex,
      match,
    } = this.props

    const {
      bookId,
      chapterIndex,
    } = match.params

    const chapterText = this.state.chapterText
    if (chapterText === '')
      return undefined
    else {
      return {
        nextChapterUrl: this.getNextChapterUrl(bookId, chapterIndex),
        text: chapterText,
        title: getChapterByIndex(bookId, chapterIndex)
      }
    }
  }

  scrollToBookmarkPosition = (prevChapterText) => {
    const {
          chapterText,
          bookmark,
        } = this.state
    const didDownloadText = prevChapterText === ""
                            && chapterText !== prevChapterText
                            && bookmark

    if (didDownloadText)
      scrollByFraction(bookmark.fraction)
  }

  fetchChapterIfPageWillChange = (nextProps) => {
    if (this.pageWillChange(nextProps)) {
      this.setState({chapterText: ""})
      this.fetchChapter(nextProps)
    }
  }

  render() {
    const pageData = this.mapStateToViewProps()
    return <ChapterPageView pageData={pageData} />
  }

  componentWillReceiveProps(nextProps) {
    this.fetchChapterIfPageWillChange(nextProps)
  }

  componentDidMount() {
    this.fetchChapter(this.props)
  }

  componentDidUpdate(prevProps, prevState) {
    this.scrollToBookmarkPosition(prevState.chapterText)
  }
}

ChapterPage.propTypes = {
  getChapterByIndex: PropTypes.func.isRequired,
  match: PropTypes.object,
  location: PropTypes.object,
}

export default ChapterPage;
