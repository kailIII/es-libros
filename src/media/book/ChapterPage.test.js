import React from "react";
import renderer from "react-test-renderer";
import ShallowRenderer from "react-test-renderer/shallow";
import { BrowserRouter as Router, MemoryRouter, Route } from "react-router-dom";

import ChapterPage from "./ChapterPage.js";
import ChapterPageRoute from "./ChapterPageRoute.js";
import ChapterPageView from "./ChapterPageView.js";

jest.mock("../../server_data/PreloadedState", () => {
  return {
    bookIndex: [
    {
      "chapters": [
        "Capitulo 1",
        "Capitulo 2"
      ]
    }
  ],
  }
})
jest.mock("../../storage/storage.js");
jest.mock("../../dom/Scroll.js", () => {
  return {
    scrollByFraction: jest.fn()
  };
});

const mockedStorage = require("../../storage/storage.js");
const mockedScroll = require("../../dom/Scroll.js").scrollByFraction;

afterEach(() => {
  mockedScroll.mockClear();
  mockedStorage.__clear();
});

const mockSuccessfulResponse = response => {
  require("superagent").__execSuccess(response);
};

const newChapterPage = () => {
  return new ChapterPage(props);
};

const expectTreeToMatchSnapshot = component => {
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
};

it("Renders progress wheel until successful response is received", () => {
  const props = {
    getChapterByIndex: () => "Chapter 1",
    match: {
      params: {
        bookId: "0",
        chapterIndex: "0"
      }
    },
    location: {}
  };
  //initial
  const component = renderer.create(
    <Router>
      <ChapterPage {...props} />
    </Router>
  );
  expectTreeToMatchSnapshot(component);

  //after response
  const successfulResponse = {
    text: "This is the page content."
  };
  mockSuccessfulResponse(successfulResponse);
  expectTreeToMatchSnapshot(component);
});

it("scrolls to bookmark position after text is received if has bookmark in local storage and querystring has bookmark=1", () => {
  const props = {
    getChapterByIndex: () => "Chapter 1",
    match: {
      params: {
        bookId: "0",
        chapterIndex: "0"
      }
    },
    location: {
      search: "?bookmark=1"
    }
  };
  const component = renderer.create(
    <Router>
      <ChapterPage {...props} />
    </Router>
  );

  mockedStorage.addBookmark(0, 0, 0.5);
  //after response
  const successfulResponse = {
    text: "This is the page content."
  };
  mockSuccessfulResponse(successfulResponse);
  expect(mockedScroll.mock.calls).toEqual([[0.5]]);
});

it("does not scroll if querystring does not have bookmark=1", () => {
  const props = {
    getChapterByIndex: () => "Chapter 1",
    match: {
      params: {
        bookId: "0",
        chapterIndex: "0"
      }
    },
    location: {}
  };
  const component = renderer.create(
    <Router>
      <ChapterPage {...props} />
    </Router>
  );

  mockedStorage.addBookmark(0, 0, 0.5);
  //after response
  const successfulResponse = {
    text: "This is the page content."
  };
  mockSuccessfulResponse(successfulResponse);
  expect(mockedScroll.mock.calls).toEqual([]);
});

it("clears bookmark from state when moving to next page", () => {
  const component = renderer.create(
    <MemoryRouter initialEntries={["/book/0/0?bookmark=1"]}>
      <ChapterPageRoute />
    </MemoryRouter>
  );

  mockedStorage.addBookmark(0, 0, 0.5);
  //after response
  const successfulResponse = {
    text: "This is the page content."
  };
  mockSuccessfulResponse(successfulResponse);

  expect(component.toJSON()).toMatchSnapshot();
  expect(mockedScroll.mock.calls).toEqual([[0.5]]);

  //click next chapter button
  const linkComponent = component.toJSON().children[1];
  linkComponent.props.onClick({
    defaultPrevented: false,
    button: 0,
    preventDefault: () => {}
  });

  // after next chapter response
  const successful2ndResponse = {
    text: "This is the next page content."
  };
  mockSuccessfulResponse(successful2ndResponse);

  expect(component.toJSON()).toMatchSnapshot();
  //number of scrolls must remain the same
  expect(mockedScroll.mock.calls).toEqual([[0.5]]);
});
