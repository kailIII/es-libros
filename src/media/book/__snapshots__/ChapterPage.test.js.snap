exports[`test Renders progress wheel until successful response is received 1`] = `
<div
  className="loader"
  style={Object {}} />
`;

exports[`test Renders progress wheel until successful response is received 2`] = `
<div>
  <div
    className="container-markdown">
    <h2
      className="book-markdown">
      Chapter 1
    </h2>
    <p
      className="book-markdown">
      This is the page content.
    </p>
  </div>
  <a
    className="next-chapter"
    href="/book/0/1"
    onClick={[Function]}>
    Siguiente Capitulo
  </a>
</div>
`;

exports[`test clears bookmark from state when moving to next page 1`] = `
<div>
  <div
    className="container-markdown">
    <h2
      className="book-markdown">
      Capitulo 1
    </h2>
    <p
      className="book-markdown">
      This is the page content.
    </p>
  </div>
  <a
    className="next-chapter"
    href="/book/0/1"
    onClick={[Function]}>
    Siguiente Capitulo
  </a>
</div>
`;

exports[`test clears bookmark from state when moving to next page 2`] = `
<div>
  <div
    className="container-markdown">
    <h2
      className="book-markdown">
      Capitulo 2
    </h2>
    <p
      className="book-markdown">
      This is the next page content.
    </p>
  </div>
</div>
`;
